src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,228 @@
+        throw;
+      }
+    }
+
+    public ValueTask DisposeAsync()
+    {
+      if (this.disposed)
+      {
+        return default;
+      }
+
+      this.disposed = true;
+
+      this.cancellationTokenSource.Cancel();
+      this.cancellationTokenSource.Dispose();
+
+      return this.resourceReaperContainer?.DisposeAsync() ?? default;
+    }
+
+    private async Task StartAsync(CancellationToken ct = default)
+    {
+      var ryukInitializedTaskSource = new TaskCompletionSource<bool>();
+
+      await this.resourceReaperContainer.StartAsync(ct)
+        .ConfigureAwait(false);
+
+      _ = this.MaintainRyukConnection(ryukInitializedTaskSource, ct);
+
+      await ryukInitializedTaskSource.Task
+        .ConfigureAwait(false);
+    }
+
+    private async Task MaintainRyukConnection(TaskCompletionSource<bool> ryukInitializedTaskSource, CancellationToken ct = default)
+    {
+      using (var initTimeoutCancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(10)))
+      {
+        using (var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.cancellationTokenSource.Token, initTimeoutCancellationTokenSource.Token, ct))
+        {
+          var endpoint = new IPEndPoint(IPAddress.Loopback, this.resourceReaperContainer.GetMappedPublicPort(RyukPort));
+
+          using (var tcpClient = new TcpClient())
+          {
+            while (!linkedCancellationTokenSource.IsCancellationRequested)
+            {
+              try
+              {
+                tcpClient.Connect(endpoint);
+
+                var stream = tcpClient.GetStream();
+
+                var filter = $"label={ResourceReaperSessionLabel}={this.SessionId:D}\n";
+
+                var sendBytes = Encoding.ASCII.GetBytes(filter);
+
+                var readBytes = new byte[4];
+
+                if (!ryukInitializedTaskSource.Task.IsCompleted)
+                {
+                  bool hasAcknowledge;
+
+                  await stream.WriteAsync(sendBytes, 0, sendBytes.Length, linkedCancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+
+                  await stream.FlushAsync(linkedCancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+
+                  do
+                  {
+                    _ = await stream.ReadAsync(readBytes, 0, readBytes.Length, linkedCancellationTokenSource.Token)
+                      .ConfigureAwait(false);
+                    hasAcknowledge = "ack\n".Equals(Encoding.ASCII.GetString(readBytes), StringComparison.OrdinalIgnoreCase);
+                  }
+                  while (!hasAcknowledge);
+
+                  ryukInitializedTaskSource.SetResult(true);
+                }
+
+                if (!linkedCancellationTokenSource.IsCancellationRequested)
+                {
+                  // Keep the connection to Ryuk up.
+                  _ = await stream.ReadAsync(readBytes, 0, readBytes.Length, default)
```

Comment:
The `cancellationToken` parameter on this `ReadAsync` should be `linkedCancellationTokenSource.Token`.

Should we also wrap this in a while loop in case future Ryuk containers are more noisy?

Reply:
> The cancellationToken parameter on this ReadAsync should be linkedCancellationTokenSource.Token.

Are you sure? This will cancel `ReadAsync` after 10 seconds (`initTimeoutCancellationTokenSource`).

> Should we also wrap this in a while loop in case future Ryuk containers are more noisy?

👍

Reply:
I changed that

Reply:
Oh, you are right. We should use the `ct` parameter here and in the while loops in this method. The `linkedCancellationTokenSource` should only be used in the initialization part of this method.

Reply:
I replaced it with the `ct`.

Reply:
Nice :+1: 
What do you think about the token in the loops? I think they should use `ct`, too.

Reply:
Yea, ofc - I missed that...